File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.cpp
class Solution {
public:
    void generateHappyStr(int n, string currentStr, vector<string>& happyStr) {
        if (currentStr.size() == n) {
            happyStr.push_back(currentStr);
            return;
        }

        for (char ch = 'a'; ch <= 'c'; ch++) {
            if (currentStr.size() > 0 && currentStr.back() == ch)
                continue;
            generateHappyStr(n, currentStr + ch, happyStr);
        }
    }

    string getHappyString(int n, int k) {
        vector<string> happyStr;
        generateHappyStr(n, "", happyStr);

        if (happyStr.size() < k)
            return "";

        sort(happyStr.begin(), happyStr.end());

        return happyStr[k - 1];
    }
};

File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.go
import (
	"sort"
)

func getHappyString(n int, k int) string {
	var happyStr []string
	generateHappyStr(n, "", &happyStr)
	if len(happyStr) < k {
		return ""
	}
	sort.Strings(happyStr)
	return happyStr[k-1]
}

func generateHappyStr(n int, currentStr string, happyStr *[]string) {
	if len(currentStr) == n {
		*happyStr = append(*happyStr, currentStr)
		return
	}

	for ch := 'a'; ch <= 'c'; ch++ {
		if len(currentStr) > 0 && rune(currentStr[len(currentStr)-1]) == ch {
			continue
		}

		generateHappyStr(n, currentStr+string(ch), happyStr)
	}
}